src/collapse/perf.rs

```diff
@@ -154,29 +188,96 @@ impl Collapse for Folder {
+                break;
+            }
+            if line.starts_with('#') {
+                continue;
+            }
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else if self.in_event {
+                self.on_stack_line(line)
+            } else {
+                self.on_event_line(line)
+            }
         }
+        Ok(())
     }
-}
 
-impl Folder {
-    fn on_line(&mut self, line: &str) {
-        if !self.in_event {
-            self.on_event_line(line)
-        } else {
-            self.on_stack_line(line)
-        }
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        debug_assert!(self.occurrences.is_concurrent());
+        debug_assert!(self.nthreads > 1);
+
+        let mut buf = Vec::with_capacity(CAPACITY_INPUT_BUFFER);
+        reader.read_to_end(&mut buf)?;
+
+        let mut input = Input::new(
+            buf,
+            self.nthreads,
+            Self::identify_stack_locations(&mut self.event_filter_state),
+        )?;
+
+        crossbeam::thread::scope(|scope| {
+            let mut handles = Vec::with_capacity(input.nthreads());
+            let (sender, receiver) = crossbeam::channel::bounded(input.nthreads());
+            for chunk in input.chunks() {
+                let event_filter_state = self.event_filter_state.clone();
+                let occurrences = self.occurrences.clone();
+
+                let annotate_jit = self.annotate_jit;
+                let annotate_kernel = self.annotate_kernel;
+                let include_addrs = self.include_addrs;
+                let include_pid = self.include_pid;
+                let include_tid = self.include_tid;
+
+                let sender = sender.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        // state
+                        cache_line: Vec::default(),
+                        event_filter_state,
+                        in_event: false,
+                        occurrences,
+                        skip_stack: false,
+                        stack: VecDeque::default(),
+                        pname: String::new(),
+
+                        // options
+                        annotate_jit,
+                        annotate_kernel,
+                        include_addrs,
+                        include_pid,
+                        include_tid,
+                        nthreads: 1,
+                    };
+                    let result = folder.collapse_single_threaded(chunk);
+                    sender.send(result).unwrap();
+                });
+                handles.push(handle);
+            }
+            for handle in handles {
+                receiver.recv().unwrap()?;
+                handle.join().unwrap();
+            }
+            Ok::<_, io::Error>(())
+        })
+        .unwrap()?;
```

Comment:
Same trick here as in dtrace: Remove `?;` and the `Ok(())` below. Should also let you remove the type annotation above.